cpu/rp2040/periph/uart.c

```diff
@@ -0,0 +1,211 @@
+ *
+ * @file
+ * @brief       UART driver implementation for the RP2040
+ *
+ * @author      Franz Freitag <franz.freitag@st.ovgu.de>
+ * @author      Justus Krebs <justus.krebs@st.ovgu.de>
+ * @author      Nick Weiler <nick.weiler@st.ovgu.de>
+ * @author      Marian Buschsieweke <marian.buschsieweke@ovgu.de>
+ *
+ * @}
+ */
+
+#include "board.h"
+#include "bitarithm.h"
+#include "irq.h"
+#include "mutex.h"
+#include "periph/gpio.h"
+#include "periph/uart.h"
+#include "periph_cpu.h"
+#include "periph_conf.h"
+#include "io_reg.h"
+
+#include "assert.h"
+
+static uart_isr_ctx_t ctx[UART_NUMOF];
+static mutex_t tx_lock = MUTEX_INIT_LOCKED;
+
+void _irq_enable(uart_t uart)
+{
+    UART0_Type *dev = uart_config[uart].dev;
+    unsigned irq_state = irq_disable();
+    /* register callback */
+    uint32_t irqs = UART0_UARTIMSC_TXIM_Msk;
+    if (ctx[uart].rx_cb) {
+        irqs |= UART0_UARTIMSC_RXIM_Msk;
+    }
+    dev->UARTIMSC.reg = irqs;
+
+    irq_restore(irq_state);
+    NVIC_EnableIRQ(uart_config[uart].irqn);
+}
+
+void _set_symbolrate(uart_t uart, uint32_t baud)
+{
+    assert(baud != 0);
+    UART0_Type *dev = uart_config[uart].dev;
+    uint32_t baud_rate_div = (8 * CLOCK_CLKPERI / baud);
+    uint32_t baud_ibrd = baud_rate_div >> 7;
+    uint32_t baud_fbrd;
+
+    if (baud_ibrd == 0) {
+        baud_ibrd = 1;
+        baud_fbrd = 0;
+    }
+    else if (baud_ibrd >= 65535) {
+        baud_ibrd = 65535;
+        baud_fbrd = 0;
+    }
+    else {
+        baud_fbrd = ((baud_rate_div & 0x7f) + 1) / 2;
+    }
+
+    io_reg_atomic_set(&(dev->UARTIBRD.reg), baud_ibrd & UART0_UARTIBRD_BAUD_DIVINT_Msk);
+    io_reg_atomic_set(&(dev->UARTFBRD.reg), baud_fbrd & UART0_UARTFBRD_BAUD_DIVFRAC_Msk);
+
+    io_reg_atomic_set(&(dev->UARTLCR_H.reg), 0);
+}
+
+int uart_mode(uart_t uart, uart_data_bits_t data_bits, uart_parity_t uart_parity,
+              uart_stop_bits_t stop_bits)
+{
+    UART0_Type *dev = uart_config[uart].dev;
+
+    uint32_t config = (data_bits << UART0_UARTLCR_H_WLEN_Pos);
+
+    switch (uart_parity) {
+    case UART_PARITY_NONE:
+        break;
+    case UART_PARITY_EVEN:
+        config |= UART0_UARTLCR_H_EPS_Msk | UART0_UARTLCR_H_PEN_Msk;
+        break;
+    case UART_PARITY_ODD:
+        config |= UART0_UARTLCR_H_PEN_Msk;
+        break;
+    default:
+        return UART_NOMODE;
+    }
+
+    if (stop_bits == UART_STOP_BITS_2) {
+        config |= UART0_UARTLCR_H_STP2_Msk;
+    }
+
+    dev->UARTLCR_H.reg = config;
+
+    return UART_OK;
+}
+
+void uart_init_pins(uart_t uart)
+{
+    gpio_init(uart_config[uart].tx_pin, GPIO_OUT);
+    volatile gpio_io_ctrl_t *io_config_reg_tx = gpio_io_register(uart_config[uart].tx_pin);
+    io_config_reg_tx->function_select = FUNCTION_SELECT_UART;
+    if (ctx[uart].rx_cb) {
+        gpio_init(uart_config[uart].rx_pin, GPIO_IN_PU);
+        volatile gpio_io_ctrl_t *io_config_reg_rx = gpio_io_register(uart_config[uart].rx_pin);
+        io_config_reg_rx->function_select = FUNCTION_SELECT_UART;
```

Comment:
Maybe we should rather have a `gpio_init_mux(uart_config[uart].rx_pin, FUNCTION_SELECT_UART)` so drivers don't have to call `gpio_io_register()` directly?

Reply:
Oha, that is actually a potentially corrupting write, as there is no guarantee that the compiler will use a 32 bit read modify write sequence for the access. I will reconsider the API altogether, as this looks very much like a valid use of the API.

Reply:
And there already is `gpio_set_function_select()` to safely do this.